include/circt/Dialect/LLHD/IR/InsertOps.td

```diff
@@ -0,0 +1,125 @@
+
+    %vtarget = constant dense<[1,2,3]> : vector<3xi32>
+    %vslice = constant dense<[4,5]> : vector<2xi32>
+    %1 = llhd.inss %vtarget, %vslice, 0 : vector<3xi32>, vector<2xi32>
+    ```
+  }];
+
+  let arguments = (ins AnyTypeOf<[AnySignlessInteger, AnyVector]>: $target,
+                       AnyTypeOf<[AnySignlessInteger, AnyVector]>: $slice,
+                       IndexAttr: $start);
+
+  let results = (outs AnyTypeOf<[AnySignlessInteger, AnyVector]>: $result);
+
+  let assemblyFormat = [{
+    $target `,` $slice `,` $start attr-dict `:` type($target) `,` type($slice)
+  }];
+
+  let extraClassDeclaration = [{
+    unsigned getSliceSize() {
+      Type sliceType = slice().getType();
+      if (auto vec = sliceType.dyn_cast<VectorType>())
+        return vec.getNumElements();
+      return sliceType.getIntOrFloatBitWidth();
+    }
+
+    unsigned getTargetSize() {
+      Type targetType = target().getType();
+      if (auto vec = targetType.dyn_cast<VectorType>())
+        return vec.getNumElements();
+      return targetType.getIntOrFloatBitWidth();
+    }
+  }];
+}
+
+def LLHD_InsfOp : LLHD_Op<"insf", [
+    NoSideEffect,
+    AllTypesMatch<["target", "result"]>,
+    PredOpTrait<"'index' has to be smaller than the 'target' size",
+      CPred<"$index.cast<IntegerAttr>().getInt() < this->getTargetSize()">>,
+    TypesMatchWith<"'element' type has to match type at 'index' of 'target'",
+      "target", "element",
+      "this->getElementTypeAtIndex($index.cast<IntegerAttr>().getInt())">
+  ]> {
+  let summary = "Insert an element into a vector or tuple.";
+  let description = [{
+    The `llhd.insf` operation allows insertion of an element represented by
+    the `$element` operand into the `$target` operand. The `$index`
+    attribute defines the index where to insert the element. The return type
+    is the same as `$target`. Note that the `$target` is not changed, but a
+    new value with the element inserted is returned.
+
+    Example:
+
+    ```mlir
+    %target = constant dense<[1,2,3]> : vector<3xi8>
+    %element = llhd.const 2 : i8
+    %0 = llhd.insf %target, %element, 0 : vector<3xi8>, i8
+
+    %tuptarget = llhd.tuple %element, %target : tuple<i8, vector<3xi8>
+    %newelement = llhd.const 4 : i8
+    %1 = llhd.insf %tuptarget, %newelement, 0 : tuple<i8, vector<3xi8>>, i8
+    ```
+  }];
+
+  let arguments = (ins AnyTypeOf<[AnyVector, AnyTuple]>: $target,
+                       AnyType: $element,
+                       IndexAttr: $index);
+
+  let results = (outs AnyTypeOf<[AnyVector, AnyTuple]>: $result);
+
+  let assemblyFormat = [{
+    $target `,` $element `,` $index attr-dict `:`
+    type($target) `,` type($element)
+  }];
+
+  let extraClassDeclaration = [{
+    int64_t getTargetSize() {
+      Type targetType = target().getType();
+      if (auto vec = targetType.dyn_cast<VectorType>())
+        return vec.getNumElements();
+      return targetType.dyn_cast<TupleType>().size();
```

Comment:
This (and the one in the next method down) should be 'cast' instead of dyn_cast given that you unconditionally use it.